### ISSUE TITLE
Import namespace collision with project module `token.py`

### ISSUE THREAD

**Issue by ejmg** [pythonic-emacs/anaconda-mode#318]
Hi there,

Thank you for making such a great emacs mode, I use it all the time.

I believe I encountered a possible error with how anaconda-mode handles completion candidates, bear with me.

In my project, `loxy`, I have the following structure:

```
.
├── loxy
│   ├── __init__.py
│   ├── loxy.py
│   ├── __pycache__
│   │   ├── <TRIMMED>
│   ├── scanner.py
│   ├── token.py
│   └── token_type.py
├── loxy.egg-info
│   ├── <TRIMMED>
└── setup.py
```

After doing some work on my project and exiting out, I have the following import statement across some of my files:
```
from loxy.token import token
# or
from loxy.token_type import TokenType
# etc
```

Re-opening emacs later yielded me the following error, initially:
```
Traceback (most recent call last):
  File "<string>", line 49, in <module>
  File "/home/spook/.pyenv/versions/lox/lib/python3.6/site-packages/setuptools/__init__.py", line 12, in <module>
    import setuptools.version
  File "/home/spook/.pyenv/versions/lox/lib/python3.6/site-packages/setuptools/version.py", line 1, in <module>
    import pkg_resources
  File "/home/spook/.pyenv/versions/lox/lib/python3.6/site-packages/pkg_resources/__init__.py", line 26, in <module>
    import zipfile
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/zipfile.py", line 20, in <module>
    import dummy_threading as threading
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/dummy_threading.py", line 45, in <module>
    import threading
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/threading.py", line 7, in <module>
    from traceback import format_exc as _format_exc
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/traceback.py", line 5, in <module>
    import linecache
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/linecache.py", line 11, in <module>
    import tokenize
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/tokenize.py", line 35, in <module>
    from token import *
  File "/home/spook/code/learning/loxy/loxy/token.py", line 1, in <module>
    from loxy.token import TokenType
  File "/home/spook/code/learning/loxy/loxy/loxy.py", line 5, in <module>
    from loxy.scanner import Scanner
ModuleNotFoundError: No module named 'loxy.scanner'; 'loxy' is not a package

```

All completion behavior is gone. This keeps getting spammed _*over and over and over*_ again in the `*anaconda-mode*` buffer.

So in my attempt to debug this behavior, I move to relative imports with the format of:
```
from .loxy.token import token
# etc
```

And now I get the following error, over and over again as before:
```
Traceback (most recent call last):
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/zipfile.py", line 18, in <module>
    import threading
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/threading.py", line 7, in <module>
    from traceback import format_exc as _format_exc
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/traceback.py", line 5, in <module>
    import linecache
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/linecache.py", line 11, in <module>
    import tokenize
  File "/home/spook/.pyenv/versions/3.6.5/lib/python3.6/tokenize.py", line 35, in <module>
    from token import *
  File "/home/spook/code/learning/loxy/loxy/token.py", line 1, in <module>
    from loxy.token import TokenType
  File "/home/spook/code/learning/loxy/loxy/loxy.py", line 5, in <module>
    from .scanner import Scanner
ImportError: attempted relative import with no known parent package
```
After quite a bit of debugging and going on the python irc, I simply rename `token.py` to `token_.py`. *Errors are completely gone*.

I am wondering what exactly what was happening here with anaconda mode and what could possibly be done to prevent this behavior?

**Comment by valignatev** [comment]
In Python, there's built-in library called `token`: https://docs.python.org/3.7/library/token.html
Also, `jedi` library (which is used in anaconda-mode under the hood) has `token` as well. It seems like it overtakes your `token.py` file and gets confused with the path. Maybe your package is not installed in dev mode correctly? What do you use for this? `pip install -e .`?

**Comment by ejmg** [comment]
Hi @valignatev, thanks for getting back to me.

Yeah, I assumed this was a name shadowing error of some sort given my "solution."

To answer you question, I install my project locally using `pip install -e .` as you guessed.

**Comment by valignatev** [comment]
Interesting if it still fails if you rename it back to "token". Maybe I saw something similar with built-in email module and my (or Django's) module with the same name

**Comment by CeleritasCelery** [comment]
@ejmg have you been able to resolve this collision?

**Comment by ejmg** [comment]
I am just now seeing this question, so sorry for missing it.

IIRC, what happened was that when I initiated anaconda-mode from within the module at hand (i.e. token.py), it broke, but when I initiated anaconda-mode for the project at a file higher in the project hierarchy, it didn't break. This ties right back into namespace stuff, I figure.

In other words, though, no, I didn't really "resolve" it. I just worked around it :slightly_frowning_face:
